File: test/python-tests/test_memory_management.py
#!/usr/bin/env python3
"""
测试Python的内存管理特性：__slots__、weakref、gc模块
"""

import sys
import gc
import weakref
from collections import defaultdict
import time
import tracemalloc

class RegularClass:
    """普通类 - 用于对比内存使用"""
    
    def __init__(self, name, value):
        self.name = name
        self.value = value
        self.data = list(range(100))  # 一些数据
    
    def __repr__(self):
        return f"RegularClass(name='{self.name}', value={self.value})"

class SlottedClass:
    """使用__slots__的类 - 内存优化"""
    
    __slots__ = ['name', 'value', 'data']  # 只允许这些属性
    
    def __init__(self, name, value):
        self.name = name
        self.value = value
        self.data = list(range(100))  # 一些数据
    
    def __repr__(self):
        return f"SlottedClass(name='{self.name}', value={self.value})"

class SlottedWithInheritance(SlottedClass):
    """继承自带slots的类"""
    
    __slots__ = ['extra_field']  # 添加新的slot
    
    def __init__(self, name, value, extra_field=None):
        super().__init__(name, value)
        self.extra_field = extra_field

def test_slots_memory_usage():
    """测试__slots__的内存使用优势"""
    print("=== Testing __slots__ Memory Usage ===")
    
    # 创建大量实例进行对比
    num_instances = 10000
    
    print(f"Creating {num_instances} instances of each class...")
    
    # 测试普通类
    start_time = time.time()
    regular_instances = [RegularClass(f"obj_{i}", i) for i in range(num_instances)]
    regular_time = time.time() - start_time
    
    # 测试带slots的类
    start_time = time.time()
    slotted_instances = [SlottedClass(f"obj_{i}", i) for i in range(num_instances)]
    slotted_time = time.time() - start_time
    
    # 计算内存使用（近似）
    regular_size = sys.getsizeof(regular_instances[0]) * num_instances
    slotted_size = sys.getsizeof(slotted_instances[0]) * num_instances
    
    print(f"Regular class:")
    print(f"  Creation time: {regular_time:.3f}s")
    print(f"  Approximate memory: {regular_size:,} bytes")
    print(f"  Instance size: {sys.getsizeof(regular_instances[0])} bytes")
    
    print(f"Slotted class:")
    print(f"  Creation time: {slotted_time:.3f}s")
    print(f"  Approximate memory: {slotted_size:,} bytes")
    print(f"  Instance size: {sys.getsizeof(slotted_instances[0])} bytes")
    
    print(f"Memory savings: {((regular_size - slotted_size) / regular_size * 100):.1f}%")
    
    # 测试属性访问限制
    print("\n--- Testing Slot Restrictions ---")
    
    slotted_obj = SlottedClass("test", 42)
    print(f"Allowed attributes: {slotted_obj.__slots__}")
    
    # 正常访问
    print(f"name: {slotted_obj.name}")
    print(f"value: {slotted_obj.value}")
    
    # 尝试添加未在slots中定义的属性
    try:
        slotted_obj.new_attribute = "should_fail"
        print("ERROR: Should not be able to add new attributes to slotted class")
    except AttributeError as e:
        print(f"✓ Correctly prevented new attribute: {e}")

def test_slots_inheritance():
    """测试带slots类的继承"""
    print("\n=== Testing Slots Inheritance ===")
    
    # 创建继承实例
    inherited = SlottedWithInheritance("test", 100, "extra_value")
    
    print(f"Inherited slots: {inherited.__slots__}")
    print(f"Name: {inherited.name} (from parent)")
    print(f"Value: {inherited.value} (from parent)")
    print(f"Extra field: {inherited.extra_field} (from child)")
    
    # 测试内存使用
    print(f"Instance size: {sys.getsizeof(inherited)} bytes")
    
    # 测试新属性限制
    try:
        inherited.another_new_attr = "should_fail"
        print("ERROR: Should not be able to add new attributes")
    except AttributeError as e:
        print(f"✓ Correctly prevented new attribute in inherited class: {e}")

def test_weak_references():
    """测试弱引用"""
    print("\n=== Testing Weak References ===")
    
    class DataObject:
        """测试对象"""
        def __init__(self, value):
            self.value = value
        
        def __repr__(self):
            return f"DataObject({self.value})"
    
    # 创建对象和弱引用
    obj1 = DataObject(100)
    obj2 = DataObject(200)
    
    # 创建弱引用
    weak_ref1 = weakref.ref(obj1)
    weak_ref2 = weakref.ref(obj2)
    
    print(f"Original objects: {obj1}, {obj2}")
    print(f"Weak references: {weak_ref1()}, {weak_ref2()}")
    
    # 测试弱引用在对象存在时的行为
    print(f"Weak reference is alive: {weak_ref1() is not None}")
    print(f"Weak reference value: {weak_ref1().value if weak_ref1() else 'Object deleted'}")
    
    # 删除强引用
    del obj1
    
    # 强制垃圾回收（可选）
    gc.collect()
    
    print(f"After deleting obj1:")
    print(f"Weak reference 1 is alive: {weak_ref1() is not None}")
    print(f"Weak reference 2 is alive: {weak_ref2() is not None}")
    
    # 测试WeakValueDictionary
    print("\n--- Testing WeakValueDictionary ---")
    
    weak_dict = weakref.WeakValueDictionary()
    
    # 创建临时对象
    temp_obj1 = DataObject(300)
    temp_obj2 = DataObject(400)
    
    # 存储到弱引用字典
    weak_dict['obj1'] = temp_obj1
    weak_dict['obj2'] = temp_obj2
    
    print(f"Weak dict before deletion: {dict(weak_dict)}")
    
    # 删除临时对象
    del temp_obj1
    gc.collect()
    
    print(f"Weak dict after deleting obj1: {dict(weak_dict)}")
    
    # 测试WeakKeyDictionary
    print("\n--- Testing WeakKeyDictionary ---")
    
    weak_key_dict = weakref.WeakKeyDictionary()
    
    key1 = DataObject("key1")
    key2 = DataObject("key2")
    
    weak_key_dict[key1] = "value1"
    weak_key_dict[key2] = "value2"
    
    print(f"Weak key dict before deletion: {dict(weak_key_dict)}")
    
    del key1
    gc.collect()
    
    print(f"Weak key dict after deleting key1: {dict(weak_key_dict)}")

def test_garbage_collection():
    """测试垃圾回收"""
    print("\n=== Testing Garbage Collection ===")
    
    # 获取当前GC状态
    print(f"GC is enabled: {gc.isenabled()}")
    print(f"GC thresholds: {gc.get_threshold()}")
    print(f"Current GC count: {gc.get_count()}")
    
    # 创建一些循环引用
    class Node:
        def __init__(self, value):
            self.value = value
            self.next = None
        
        def __repr__(self):
            return f"Node({self.value})"
    
    # 创建循环引用
    node1 = Node(1)
    node2 = Node(2)
    node3 = Node(3)
    
    node1.next = node2
    node2.next = node3
    node3.next = node1  # 循环引用
    
    print(f"\nCreated circular reference: {node1} -> {node2} -> {node3} -> {node1}")
    
    # 手动禁用GC来观察引用计数
    gc.disable()
    print(f"GC disabled: {not gc.isenabled()}")
    
    # 删除引用
    del node1
    del node2
    del node3
    
    print("Deleted all references to nodes with circular reference")
    
    # 强制垃圾回收
    collected = gc.collect()
    print(f"Manually collected {collected} objects")
    
    # 重新启用GC
    gc.enable()
    print(f"GC re-enabled: {gc.isenabled()}")

def test_memory_profiling():
    """测试内存分析"""
    print("\n=== Testing Memory Profiling ===")
    
    # 启动内存跟踪
    tracemalloc.start()
    
    # 记录当前内存使用
    snapshot1 = tracemalloc.take_snapshot()
    
    # 创建一些对象
    large_list = [i for i in range(100000)]
    
    # 记录内存使用变化
    snapshot2 = tracemalloc.take_snapshot()
    
    # 计算内存差异
    top_stats = snapshot2.compare_to(snapshot1, 'lineno')
    
    print("Top memory allocations:")
    for stat in top_stats[:5]:
        print(f"  {stat}")
    
    # 获取当前内存使用统计
    current, peak = tracemalloc.get_traced_memory()
    print(f"\nCurrent memory usage: {current / 1024 / 1024:.2f} MB")
    print(f"Peak memory usage: {peak / 1024 / 1024:.2f} MB")
    
    # 停止内存跟踪
    tracemalloc.stop()

def test_object_lifecycle():
    """测试对象生命周期管理"""
    print("\n=== Testing Object Lifecycle ===")
    
    class ManagedObject:
        """带生命周期管理的对象"""
        
        _instances = weakref.WeakSet()
        
        def __init__(self, name):
            self.name = name
            self._instances.add(self)
            print(f"Created: {self}")
        
        def __del__(self):
            print(f"Destroyed: {self}")
        
        def __repr__(self):
            return f"ManagedObject({self.name})"
        
        @classmethod
        def get_active_instances(cls):
            """获取当前活动的实例"""
            return list(cls._instances)
        
        @classmethod
        def cleanup_expired_instances(cls):
            """清理已失效的实例"""
            active = cls.get_active_instances()
            print(f"Active instances: {len(active)}")
            return active
    
    # 创建一些对象
    obj1 = ManagedObject("first")
    obj2 = ManagedObject("second")
    obj3 = ManagedObject("third")
    
    print(f"Active instances after creation: {len(ManagedObject.get_active_instances())}")
    
    # 删除一些对象
    del obj1
    gc.collect()  # 强制垃圾回收
    
    print(f"Active instances after deletion: {len(ManagedObject.get_active_instances())}")
    
    # 使用弱引用跟踪
    weak_set = weakref.WeakSet()
    weak_set.add(obj2)
    weak_set.add(obj3)
    
    print(f"Objects in weak set: {len(weak_set)}")
    
    del obj2
    gc.collect()
    
    print(f"Objects in weak set after deletion: {len(weak_set)}")

def test_memory_optimization_patterns():
    """测试内存优化模式"""
    print("\n=== Testing Memory Optimization Patterns ===")
    
    # 模式1: 使用slots减少内存占用
    class RegularPoint:
        def __init__(self, x, y):
            self.x = x
            self.y = y
    
    class SlottedPoint:
        __slots__ = ['x', 'y']
        
        def __init__(self, x, y):
            self.x = x
            self.y = y
    
    # 创建大量实例进行对比
    num_points = 100000
    
    regular_points = [RegularPoint(i, i) for i in range(num_points)]
    slotted_points = [SlottedPoint(i, i) for i in range(num_points)]
    
    regular_size = sum(sys.getsizeof(p) for p in regular_points[:100]) / 100
    slotted_size = sum(sys.getsizeof(p) for p in slotted_points[:100]) / 100
    
    print(f"Average size per RegularPoint: {regular_size} bytes")
    print(f"Average size per SlottedPoint: {slotted_size} bytes")
    print(f"Memory savings with slots: {((regular_size - slotted_size) / regular_size * 100):.1f}%")
    
    # 模式2: 对象池模式
    class ObjectPool:
        """简单的对象池实现"""
        
        def __init__(self, factory_func, max_size=100):
            self.factory_func = factory_func
            self.max_size = max_size
            self._pool = []
            # 用于跟踪活动对象的 id（支持不可弱引用且不可哈希的对象，如 dict）
            self._active_ids = set()
        
        def get(self):
            """从池中获取对象"""
            if self._pool:
                obj = self._pool.pop()
            else:
                obj = self.factory_func()
            
            self._active_ids.add(id(obj))
            return obj
        
        def put(self, obj):
            """将对象返回池中"""
            if len(self._pool) < self.max_size:
                self._pool.append(obj)
            self._active_ids.discard(id(obj))
        
        @property
        def pool_size(self):
            return len(self._pool)
        
        @property
        def active_count(self):
            return len(self._active_ids)
    
    # 使用对象池
    def create_expensive_object():
        return {"data": list(range(1000)), "timestamp": time.time()}
    
    pool = ObjectPool(create_expensive_object, max_size=50)
    
    # 获取和释放对象
    obj1 = pool.get()
    obj2 = pool.get()
    
    print(f"\nObject pool - Active: {pool.active_count}, Pool size: {pool.pool_size}")
    
    pool.put(obj1)
    pool.put(obj2)
    
    print(f"After returning objects - Active: {pool.active_count}, Pool size: {pool.pool_size}")

def main():
    """主测试函数"""
    print("Python Memory Management: __slots__, weakref, gc Module Demonstration")
    print("=" * 80)
    
    test_slots_memory_usage()
    test_slots_inheritance()
    test_weak_references()
    test_garbage_collection()
    test_memory_profiling()
    test_object_lifecycle()
    test_memory_optimization_patterns()
    
    print("\n=== Summary ===")
    print("Key concepts demonstrated:")
    print("1. __slots__ for memory optimization and attribute restriction")
    print("2. weakref for creating weak references that don't prevent garbage collection")
    print("3. gc module for manual garbage collection control")
    print("4. Memory profiling with tracemalloc")
    print("5. Object lifecycle management with weak references")
    print("6. Memory optimization patterns including object pooling")
    print("\nBest practices:")
    print("- Use __slots__ when creating many instances with fixed attributes")
    print("- Use weakref for caches and observer patterns to avoid circular references")
    print("- Monitor memory usage in production applications")
    print("- Implement proper cleanup in __del__ methods")

if __name__ == "__main__":
    main()

File: test/python-tests/test_memory_management.cpp
#include <iostream>
#include <string>
#include <vector>
#include <optional>
#include <unordered_map>
#include <tuple>
#include <utility>
#include <type_traits>
#include <functional>
#include <sstream>
#include <iomanip>
#include <cmath>
#include <algorithm>
#include <typeinfo>
#include <cstring>

using namespace std;

// helpers
template<typename T> static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v);
template<typename T> static inline std::string to_str(const T& v){ std::ostringstream os; os<<v; return os.str(); }
static inline std::string to_str(const std::string& s){ return s; }
static inline std::string to_str(const char* s){ return std::string(s); }
static inline std::string to_str(double v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
static inline std::string to_str(float v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
template<typename T> static inline int to_int(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stoi(v); else return static_cast<int>(v); }
template<typename T> static inline float to_float(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stof(v); else return static_cast<float>(v); }
template<typename C> static inline int sum(const C& c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline int sum(std::initializer_list<T> c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline size_t len(const T& c){ return c.size(); }
static inline size_t len(const char* s){ return std::char_traits<char>::length(s); }
template<typename K, typename V> static inline std::string to_str(const std::unordered_map<K,V>&){ return std::string("{...}"); }
// minimal ABC base to support 'class X : public ABC'
struct ABC {};
// minimal math namespace replacement for 'math.pi'
struct __fluxus_math { static constexpr double pi = 3.141592653589793; static inline double sqrt(double x){ return std::sqrt(x); } } math;
// minimal sys module stub
namespace sys { static std::vector<std::string> argv; }
namespace asyncio { template<typename F> static inline void run(F f){ f(); } }

// vector to string conversion for printing
template<typename T>
static inline std::string vec_to_str(const std::vector<T>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        if constexpr (std::is_same_v<T, std::string>) {
            os << "'" << v[i] << "'";
        } else {
            os << v[i];
        }
    }
    os << "]";
    return os.str();
}

// nested vector to string
template<typename T>
static inline std::string vec_to_str(const std::vector<std::vector<T>>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        os << vec_to_str(v[i]);
    }
    os << "]";
    return os.str();
}

// stream operator for vectors
template<typename T>
static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v) {
    os << vec_to_str(v);
    return os;
}
template<typename K, typename V> static inline std::ostream& operator<<(std::ostream& os, const std::unordered_map<K,V>&){ os << "{...}"; return os; }
template<typename K, typename V> static inline auto py_items(const std::unordered_map<K,V>& m){ std::vector<std::pair<K,V>> v; v.reserve(m.size()); for(const auto& kv: m){ v.emplace_back(kv.first, kv.second);} return v; }


class RegularClass {
    public:
        RegularClass(std::string name, int value) {
            this->name = name;
            this->value = value;
            this->data = list(range(100))  # 一些数据;
        }
        std::string __repr__() {
    return to_str("RegularClass(name='") + to_str(this->name) + to_str("', value=") + to_str(this->value) + to_str(")");
}


    public:
        std::string name = "";
        int value = 0;
        int data = 0;
};


class SlottedClass {
    public:
        SlottedClass(std::string name, int value) {
            this->name = name;
            this->value = value;
            this->data = list(range(100))  # 一些数据;
        }
        std::string __repr__() {
    return to_str("SlottedClass(name='") + to_str(this->name) + to_str("', value=") + to_str(this->value) + to_str(")");
}


    public:
        std::string name = "";
        int value = 0;
        int data = 0;
};


class SlottedWithInheritance : public SlottedClass {
    public:
        SlottedWithInheritance(std::string name, int value, int extra_field=None) : SlottedClass(name, value) {
            this->extra_field = extra_field;
        }
        void placeholder();
    public:
        int extra_field = 0;
};


void test_slots_memory_usage() {
    bool __fluxus_exc=false;
    /* docstring */
    std::cout << "=== Testing __slots__ Memory Usage ===" << std::endl;
    auto num_instances = 10000;
    std::cout << to_str("Creating ") << to_str(num_instances) << to_str(" instances of each class...") << std::endl;
    auto start_time = time.time();
    auto regular_instances = 0;
    auto regular_time = time.time() - start_time;
    auto start_time = time.time();
    auto slotted_instances = 0;
    auto slotted_time = time.time() - start_time;
    auto regular_size = sys::getsizeof(regular_instances[0]) * num_instances;
    auto slotted_size = sys::getsizeof(slotted_instances[0]) * num_instances;
    std::cout << to_str("Regular class:") << std::endl;
    std::cout << to_str("  Creation time: ") << ([&](){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(3); os<<regular_time; return os.str(); }()) << to_str("s") << std::endl;
    std::cout << to_str("  Approximate memory: ") << to_str(regular_size) << to_str(" bytes") << std::endl;
    std::cout << to_str("  Instance size: ") << to_str(sys::getsizeof(regular_instances[0])) << to_str(" bytes") << std::endl;
    std::cout << to_str("Slotted class:") << std::endl;
    std::cout << to_str("  Creation time: ") << ([&](){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(3); os<<slotted_time; return os.str(); }()) << to_str("s") << std::endl;
    std::cout << to_str("  Approximate memory: ") << to_str(slotted_size) << to_str(" bytes") << std::endl;
    std::cout << to_str("  Instance size: ") << to_str(sys::getsizeof(slotted_instances[0])) << to_str(" bytes") << std::endl;
    std::cout << to_str("Memory savings: ") << ([&](){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1); os<<((regular_size - slotted_size) / regular_size * 100); return os.str(); }()) << to_str("%") << std::endl;
    std::cout << "\n--- Testing Slot Restrictions ---" << std::endl;
    auto slotted_obj = SlottedClass("test", 42);
    std::cout << to_str("Allowed attributes: ") << to_str(slotted_obj.__slots__) << std::endl;
    std::cout << to_str("name: ") << to_str(slotted_obj.name) << std::endl;
    std::cout << to_str("value: ") << to_str(slotted_obj.value) << std::endl;
    {
        slotted_obj.new_attribute("should_fail");
        std::cout << "ERROR: Should not be able to add new attributes to slotted class" << std::endl;
    }
    if (__fluxus_exc) { auto e = 0;
        std::cout << to_str("✓ Correctly prevented new attribute: ") << to_str(e) << std::endl;
    }
}


void test_slots_inheritance() {
    bool __fluxus_exc=false;
    /* docstring */
    std::cout << "\n=== Testing Slots Inheritance ===" << std::endl;
    auto inherited = SlottedWithInheritance("test", 100, "extra_value");
    std::cout << to_str("Inherited slots: ") << to_str(inherited.__slots__) << std::endl;
    std::cout << to_str("Name: ") << to_str(inherited.name) << to_str(" (from parent)") << std::endl;
    std::cout << to_str("Value: ") << to_str(inherited.value) << to_str(" (from parent)") << std::endl;
    std::cout << to_str("Extra field: ") << to_str(inherited.extra_field) << to_str(" (from child)") << std::endl;
    std::cout << to_str("Instance size: ") << to_str(sys::getsizeof(inherited)) << to_str(" bytes") << std::endl;
    {
        inherited.another_new_attr("should_fail");
        std::cout << "ERROR: Should not be able to add new attributes" << std::endl;
    }
    if (__fluxus_exc) { auto e = 0;
        std::cout << to_str("✓ Correctly prevented new attribute in inherited class: ") << to_str(e) << std::endl;
    }
}


void test_weak_references() {
    /* docstring */
    std::cout << "\n=== Testing Weak References ===" << std::endl;
    class DataObject { public: DataObject() = default; };
    auto obj1 = DataObject(100);
    auto obj2 = DataObject(200);
    auto weak_ref1 = weakref.ref(obj1);
    auto weak_ref2 = weakref.ref(obj2);
    std::cout << to_str("Original objects: ") << to_str(obj1) << to_str(", ") << to_str(obj2) << std::endl;
    std::cout << to_str("Weak references: ") << to_str(weak_ref1()) << to_str(", ") << to_str(weak_ref2()) << std::endl;
    std::cout << to_str("Weak reference is alive: ") << to_str(weak_ref1() is not 0) << std::endl;
    std::cout << to_str("Weak reference value: ") << to_str((weak_ref1() ? weak_ref1().value : 'Object deleted')) << std::endl;
    del obj1;
    gc.collect();
    std::cout << to_str("After deleting obj1:") << std::endl;
    std::cout << to_str("Weak reference 1 is alive: ") << to_str(weak_ref1() is not 0) << std::endl;
    std::cout << to_str("Weak reference 2 is alive: ") << to_str(weak_ref2() is not 0) << std::endl;
    std::cout << "\n--- Testing WeakValueDictionary ---" << std::endl;
    auto weak_dict = weakref.WeakValueDictionary();
    auto temp_obj1 = DataObject(300);
    auto temp_obj2 = DataObject(400);
    auto weak_dict['obj1'] = temp_obj1;
    auto weak_dict['obj2'] = temp_obj2;
    std::cout << to_str("Weak dict before deletion: ") << to_str(dict(weak_dict)) << std::endl;
    del temp_obj1;
    gc.collect();
    std::cout << to_str("Weak dict after deleting obj1: ") << to_str(dict(weak_dict)) << std::endl;
    std::cout << "\n--- Testing WeakKeyDictionary ---" << std::endl;
    auto weak_key_dict = weakref.WeakKeyDictionary();
    auto key1 = DataObject("key1");
    auto key2 = DataObject("key2");
    auto weak_key_dict[key1] = "value1";
    auto weak_key_dict[key2] = "value2";
    std::cout << to_str("Weak key dict before deletion: ") << to_str(dict(weak_key_dict)) << std::endl;
    del key1;
    gc.collect();
    std::cout << to_str("Weak key dict after deleting key1: ") << to_str(dict(weak_key_dict)) << std::endl;
}


void test_garbage_collection() {
    /* docstring */
    std::cout << "\n=== Testing Garbage Collection ===" << std::endl;
    std::cout << to_str("GC is enabled: ") << to_str(gc.isenabled()) << std::endl;
    std::cout << to_str("GC thresholds: ") << to_str(gc.get_threshold()) << std::endl;
    std::cout << to_str("Current GC count: ") << to_str(gc.get_count()) << std::endl;
    class Node { public: Node() = default; };
    auto node1 = Node(1);
    auto node2 = Node(2);
    auto node3 = Node(3);
    node1.next(node2);
    node2.next(node3);
    node3.next(node1);
    std::cout << to_str("\\nCreated circular reference: ") << to_str(node1) << to_str(" -> ") << to_str(node2) << to_str(" -> ") << to_str(node3) << to_str(" -> ") << to_str(node1) << std::endl;
    gc.disable();
    std::cout << to_str("GC disabled: ") << to_str(not gc.isenabled()) << std::endl;
    del node1;
    del node2;
    del node3;
    std::cout << "Deleted all references to nodes with circular reference" << std::endl;
    auto collected = gc.collect();
    std::cout << to_str("Manually collected ") << to_str(collected) << to_str(" objects") << std::endl;
    gc.enable();
    std::cout << to_str("GC re-enabled: ") << to_str(gc.isenabled()) << std::endl;
}


void test_memory_profiling() {
    /* docstring */
    std::cout << "\n=== Testing Memory Profiling ===" << std::endl;
    tracemalloc.start();
    auto snapshot1 = tracemalloc.take_snapshot();
    auto large_list = 0;
    auto snapshot2 = tracemalloc.take_snapshot();
    auto top_stats = snapshot2.compare_to(snapshot1, 'lineno');
    std::cout << "Top memory allocations:" << std::endl;
    for (auto stat : top_stats[:5]) {
        std::cout << to_str("  ") << to_str(stat) << std::endl;
    }
    auto current, peak = tracemalloc.get_traced_memory();
    std::cout << to_str("\\nCurrent memory usage: ") << ([&](){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(2); os<<current / 1024 / 1024; return os.str(); }()) << to_str(" MB") << std::endl;
    std::cout << to_str("Peak memory usage: ") << ([&](){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(2); os<<peak / 1024 / 1024; return os.str(); }()) << to_str(" MB") << std::endl;
    tracemalloc.stop();
}


void test_object_lifecycle() {
    /* docstring */
    std::cout << "\n=== Testing Object Lifecycle ===" << std::endl;
    class ManagedObject { public: ManagedObject() = default; };
    auto obj1 = ManagedObject("first");
    auto obj2 = ManagedObject("second");
    auto obj3 = ManagedObject("third");
    std::cout << to_str("Active instances after creation: ") << to_str(len(ManagedObject.get_active_instances())) << std::endl;
    del obj1;
    gc.collect();
    std::cout << to_str("Active instances after deletion: ") << to_str(len(ManagedObject.get_active_instances())) << std::endl;
    auto weak_set = weakref.WeakSet();
    weak_set.add(obj2);
    weak_set.add(obj3);
    std::cout << to_str("Objects in weak set: ") << to_str(len(weak_set)) << std::endl;
    del obj2;
    gc.collect();
    std::cout << to_str("Objects in weak set after deletion: ") << to_str(len(weak_set)) << std::endl;
}


std::string test_memory_optimization_patterns() {
    /* docstring */
    std::cout << "\n=== Testing Memory Optimization Patterns ===" << std::endl;
    class RegularPoint { public: RegularPoint() = default; };
    class SlottedPoint { public: SlottedPoint() = default; };
    auto num_points = 100000;
    auto regular_points = 0;
    auto slotted_points = 0;
    auto regular_size = sum(std::vector<int>{});
    auto slotted_size = sum(std::vector<int>{});
    std::cout << to_str("Average size per RegularPoint: ") << to_str(regular_size) << to_str(" bytes") << std::endl;
    std::cout << to_str("Average size per SlottedPoint: ") << to_str(slotted_size) << to_str(" bytes") << std::endl;
    std::cout << to_str("Memory savings with slots: ") << ([&](){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1); os<<((regular_size - slotted_size) / regular_size * 100); return os.str(); }()) << to_str("%") << std::endl;
    class ObjectPool { public: ObjectPool() = default; };
    std::function<int()> create_expensive_object = [&]() {
        return std::unordered_map<std::string, int>{{"data", list(range(1000))}, {"timestamp", time.time()}};
    };
    auto pool = ObjectPool(create_expensive_object, max_size=50);
    auto obj1 = pool.get();
    auto obj2 = pool.get();
    std::cout << to_str("\\nObject pool - Active: ") << to_str(pool.active_count) << to_str(", Pool size: ") << to_str(pool.pool_size) << std::endl;
    pool.put(obj1);
    pool.put(obj2);
    std::cout << to_str("After returning objects - Active: ") << to_str(pool.active_count) << to_str(", Pool size: ") << to_str(pool.pool_size) << std::endl;
}


int main() {
    /* docstring */
    std::cout << "Python Memory Management: __slots__, weakref, gc Module Demonstration" << std::endl;
    std::cout << std::string(80, '=') << std::endl;
    test_slots_memory_usage();
    test_slots_inheritance();
    test_weak_references();
    test_garbage_collection();
    test_memory_profiling();
    test_object_lifecycle();
    test_memory_optimization_patterns();
    std::cout << "\n=== Summary ===" << std::endl;
    std::cout << "Key concepts demonstrated:" << std::endl;
    std::cout << 0 << std::endl;
    std::cout << 0 << std::endl;
    std::cout << 0 << std::endl;
    std::cout << "4. Memory profiling with tracemalloc" << std::endl;
    std::cout << "5. Object lifecycle management with weak references" << std::endl;
    std::cout << "6. Memory optimization patterns including object pooling" << std::endl;
    std::cout << "\nBest practices:" << std::endl;
    std::cout << "- Use __slots__ when creating many instances with fixed attributes" << std::endl;
    std::cout << 0 << std::endl;
    std::cout << "- Monitor memory usage in production applications" << std::endl;
    std::cout << "- Implement proper cleanup in __del__ methods" << std::endl;
    return 0;
}


